### PR TITLE
Interface properties over IPC

### PR DESCRIPTION
**Pull request by jacky309**

### CODE REVIEW

File: codegen/facelift/templates/ServiceIPC.template.h [hunk 1]
```diff
@@ -68,7 +72,8 @@ class {{interface}}IPCAdapter: public facelift::IPCServiceAdapter<{{interface.fu
         facelift::IPCServiceAdapter<{{interface.fullyQualifiedCppType}}>::setService(srvc);
 
         {% for property in interface.properties %}
-        {% if property.type.is_model %}
+
+            {% if property.type.is_model %}
```
Comment (weimerb) [inline codegen/facelift/templates/ServiceIPC.template.h]: Nitpick: actually I formatted it like that on purpose to make the generated code easier to read, small changes here can make it look more unstructured.

File: codegen/facelift/templates/ServiceIPC.template.h [hunk 2]
```diff
@@ -277,12 +290,15 @@ class {{interface}}IPCProxy : public facelift::IPCProxy<{{interface}}PropertyAda
...
-        qWarning() << "TODO: handle interface properties";
+        QString {{property.name}}_objectPath;
+        msg >> {{property.name}}_objectPath;
+        m_{{property.name}}Proxy.update(this, {{property.name}}_objectPath);
+        m_{{property.name}} = m_{{property.name}}Proxy.getValue();
         {% elif property.type.is_model %}
         int {{property.name}}Size;
         msg >> {{property.name}}Size;
         m_{{property.name}}.setSize({{property.name}}Size);
+
```
Comment (weimerb) [inline codegen/facelift/templates/ServiceIPC.template.h]: Then you'll probably need an empty line in the first clause, as well (should you have several interface props).

File: src/ipc/dbus/ipc-dbus-object-registry.h [hunk 3]
```diff
@@ -98,11 +98,19 @@ class DBusObjectRegistry : public facelift::ipc::ObjectRegistryPropertyAdapter
...
 
-    void unregisterObject(QString objectPath)
+    bool unregisterObject(QString objectPath, QString serviceName) override
     {
-        auto objects = m_objects.value();
-        objects.remove(objectPath);
-        m_objects = objects;
+        init();
+        if (isMaster()) {
+            Q_UNUSED(serviceName);
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus-object-registry.h]: It is actually used in the else clause.

File: src/ipc/dbus/ipc-dbus.cpp [hunk 4]
```diff
@@ -136,6 +136,13 @@ bool DBusIPCServiceAdapterBase::handleMessage(const QDBusMessage &dbusMsg, const
     return false;
 }
 
+DBusIPCServiceAdapterBase::~DBusIPCServiceAdapterBase()
+{
+    destroyed(this);
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus.cpp]: we should start adding "emit", even if it is void, it makes it easier to read.

File: src/ipc/dbus/ipc-dbus.h [hunk 5]
```diff
@@ -978,6 +977,73 @@ class DBusIPCAttachedPropertyFactory : public IPCAttachedPropertyFactoryBase
 };
 
 }
+
+
+class InterfacePropertyHandlerBase {
+public:
+
+    QString generateObjectPath(const QString& parentPath) {
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus.h]: Nitpick: "our" coding style says that the opening brace of a method should go on a new line (here and below).

File: src/ipc/dbus/ipc-dbus.h [hunk 6]
```diff
@@ -978,6 +977,73 @@ class DBusIPCAttachedPropertyFactory : public IPCAttachedPropertyFactoryBase
...
+            auto serviceAdapter = new InterfaceAdapterType(service);
+            serviceAdapter->setObjectPath(generateObjectPath(parent->objectPath()));
+            serviceAdapter->setService(service);
+            serviceAdapter->init();
+            m_serviceAdapter = std::move(std::unique_ptr<InterfaceAdapterType>(serviceAdapter));
+        }
+    }
+
+    QString objectPath() const {
+        if (m_serviceAdapter) {
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus.h]: also format, but better use tertiary operator, wouldn't QString() be better thatn ""?

File: src/ipc/dbus/ipc-dbus.cpp [hunk 7]
```diff
@@ -173,21 +173,25 @@ void DBusIPCServiceAdapterBase::init(InterfaceBase *service)
...
+            if (registry.objects().contains(objectPath())) {
+                auto serviceName = registry.objects()[objectPath()];
+                if (serviceName != m_serviceName) {
+                    m_serviceName = serviceName;
+                    bindToIPC();
                 }
-            });
-        }
+            } else {
+                m_serviceName = "";
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus.cpp]: = QString()?

**Comment by weimerb** [comment]
"Fix" as commit message is not meaningful

File: src/ipc/dbus/ipc-dbus.h [hunk 8]
```diff
@@ -978,6 +977,73 @@ class DBusIPCAttachedPropertyFactory : public IPCAttachedPropertyFactoryBase
...
+    std::unique_ptr<InterfaceAdapterType> m_serviceAdapter;
+};
+
+
+template<typename ProxyType>
+class InterfacePropertyIPCProxyHandler : public InterfacePropertyHandlerBase {
+
+public:
+
+    void update(InterfaceBase* i, const QString& objectPath) {
```
Comment (weimerb) [inline src/ipc/dbus/ipc-dbus.h]: parameter 'i' is not used.

File: src/ipc/dbus/ipc-dbus-object-registry.h (hunk 3 above)
Comment (jacky309) [reply src/ipc/dbus/ipc-dbus-object-registry.h]: done

File: src/ipc/dbus/ipc-dbus.cpp (hunk 4 above)
Comment (jacky309) [reply src/ipc/dbus/ipc-dbus.cpp]: done

File: src/ipc/dbus/ipc-dbus.h (hunk 8 above)
Comment (jacky309) [reply src/ipc/dbus/ipc-dbus.h]: fixed

**Pull request merged**